core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -466,7 +468,22 @@ private void notifyEndOfRebalanceIfNeeded(int topologyId, int rebalanceId) {
             log.debugf("Finished receiving of segments for cache %s for topology %d.", cacheName, topologyId);
             stopApplyingState();
             stateTransferManager.notifyEndOfRebalance(topologyId, rebalanceId);
-            stateRequestCompletionService.drainCompletionQueue();
+            List<? extends Future<Void>> futures = stateRequestCompletionService.drainCompletionQueue();
+            boolean interrupted = false;
+            for (Future<Void> future : futures) {
+               // These will not block since they are already completed
+               try {
+                  future.get();
+               } catch (InterruptedException e) {
+                  // This shouldn't happen as each task is complete
+                  interrupted = true;
+               } catch (ExecutionException e) {
+                  log.topologyUpdateError(topologyId, e.getCause());
```

Comment:
I had wondered about if it would always get here or not.  But I figured you would know best :)

Unfortunately we can't use the uncaught exception handler because the exception is caught, by the ThreadPoolExecutor and stored in the FutureTask.  You really need to call get on the ECS or this drainCompletionQueue.